metals/src/main/scala/scala/meta/internal/metals/Compilers.scala

```diff
@@ -1153,6 +1165,38 @@ class Compilers(
     }
   }
 
+  private def withUncachedCompiler[T](
```

Comment:
Could we merge the method at least partly with loadCompilerForTarget ? We could use `getOrDefault` instead o `computeIfAbsent` so we could add the function as a last parameter for loadCompilerForTarget.

Got a bit interested if it's possible and came up with:

```
  private def loadCompiler(
      targetId: BuildTargetIdentifier
  ): Option[PresentationCompiler] = {
    val target = buildTargets.scalaTarget(targetId)
    target.flatMap(target =>
      loadCompilerForTarget(
        target,
        { (pcKey, compiler) =>
          jcache.computeIfAbsent(pcKey, { _ => compiler() })
        },
      )
    )
  }

  private def loadCompilerUncached(
      targetId: BuildTargetIdentifier
  ): Option[PresentationCompiler] = {
    val target = buildTargets.scalaTarget(targetId)
    target.flatMap(target =>
      loadCompilerForTarget(
        target,
        { (pcKey, compiler) =>
          jcache.getOrDefault(pcKey, compiler())
        },
      )
    )
  }

  private def loadCompilerForTarget(
      scalaTarget: ScalaTarget,
      cachingAction: (
          PresentationCompilerKey,
          () => MtagsPresentationCompiler,
      ) => MtagsPresentationCompiler,
  ): Option[PresentationCompiler] = {
    val scalaVersion = scalaTarget.scalaVersion
    mtagsResolver.resolve(scalaVersion) match {
      case Some(mtags) =>
        val out = cachingAction(
          PresentationCompilerKey.ScalaBuildTarget(scalaTarget.info.getId),
         () => workDoneProgress.trackBlocking(
            s"${config.icons.sync}Loading presentation compiler"
          ) {
            ScalaLazyCompiler(scalaTarget, mtags, search)
          },
        )
        Option(out.await)
      case None =>
        scribe.warn(s"unsupported Scala ${scalaTarget.scalaVersion}")
        None
    }
  }

```

:sweat_smile:

Reply:
> `getOrDefault`

It's a java collection, so I don't think that the default will be calculated lazily.

I need to shutdown the uncached pc if it's not from cache... I don't think that your solution includes that. Most probably we can merge it somehow, I just figured it will be too awkward to be worth it.

Reply:
> It's a java collection, so I don't think that the default will be calculated lazily.

Why would we need it lazy? If it's not in cache we would just create a new one.

But anyway, it might be too awkward yeah and shutdown is not included here.

Let's at least make sure `            s"${config.icons.sync}Loading presentation compiler"` part is reused.

Reply:
```jcache.getOrDefault(pcKey, compiler())```

Reply:
Ach, right.

Reply:
I merged it after all and it ended up very similar to what you did.